File: WhipMe/Controller/Member/TopUpViewController.m
//
//  TopUpViewController.m
//  WhipMe
//
//  Created by anve on 16/11/23.
//  Copyright © 2016年 -. All rights reserved.
//

#import "TopUpViewController.h"
#import "NSString+Common.h"
#import "NSString+Validate.h"
#import "ShareEngine.h"

@interface TopUpViewController () <UITextFieldDelegate, WXApiEngineDelegate>

@property (nonatomic, strong) UIView *viewCurrent;
@property (nonatomic, strong) UITextField *textMoney;
@property (nonatomic, strong) UILabel *lblMoneyTitle;
@property (nonatomic, strong) UIButton *btnWechat, *btnSubmit;

@property (nonatomic, strong) NSString *weixin_code;
@property (nonatomic, strong) NSString *weixin_appOpenId;
@property (nonatomic, strong) NSString *weixin_unionId;

@end

@implementation TopUpViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = @"充值";
    self.view.backgroundColor = [Define kColorBackGround];
    
    [self setup];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)dealloc {
    DebugLog(@"%@",self.class);
}

- (void)setup {
    
    WEAK_SELF
    _viewCurrent = [UIView new];
    _viewCurrent.backgroundColor = [UIColor whiteColor];
    _viewCurrent.layer.cornerRadius = 4.0;
    _viewCurrent.layer.masksToBounds = true;
    [self.view addSubview:self.viewCurrent];
    [self.viewCurrent mas_updateConstraints:^(MASConstraintMaker *make) {
        make.left.and.top.equalTo(weakSelf.view).offset(10.0);
        make.right.equalTo(weakSelf.view).offset(-10.0);
        make.height.mas_equalTo(290.0);
    }];
    
    _lblMoneyTitle = [UILabel new];
    _lblMoneyTitle.backgroundColor = [UIColor clearColor];
    _lblMoneyTitle.textAlignment = NSTextAlignmentLeft;
    _lblMoneyTitle.textColor = [Define kColorGray];
    _lblMoneyTitle.font = [UIFont systemFontOfSize:16.0];
    _lblMoneyTitle.text = @"充值金额";
    [self.viewCurrent addSubview:self.lblMoneyTitle];
    [self.lblMoneyTitle mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(40.0);
        make.width.equalTo(self.viewCurrent).offset(-60.0);
        make.left.equalTo(weakSelf.view).offset(30.0);
        make.top.equalTo(self.viewCurrent);
    }];
    
    _textMoney = [UITextField new];
    _textMoney.contentVerticalAlignment = UIControlContentVerticalAlignmentCenter;
    _textMoney.backgroundColor = [UIColor clearColor];
    _textMoney.textAlignment = NSTextAlignmentLeft;
    _textMoney.font = [UIFont systemFontOfSize:22.0];
    _textMoney.keyboardType = UIKeyboardTypeDecimalPad;
    _textMoney.textColor = [Define kColorBlack];
    _textMoney.delegate = self;
    _textMoney.placeholder = @"请输入充值金额";
//    _textMoney.attributedPlaceholder = [[NSMutableAttributedString alloc] initWithString:@"请输入充值金额" attributes:@{NSFontAttributeName:[UIFont systemFontOfSize:18.0],NSForegroundColorAttributeName:[Define kColorGray]}];
    [self.viewCurrent addSubview:self.textMoney];
    [self.textMoney mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(50.0);
        make.width.equalTo(weakSelf.lblMoneyTitle.mas_width);
        make.left.equalTo(weakSelf.lblMoneyTitle.mas_left);
        make.top.equalTo(weakSelf.lblMoneyTitle.mas_bottom);
    }];
    
    UILabel *leftView = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 40.0, 50.0)];
    leftView.textAlignment = NSTextAlignmentCenter;
    leftView.textColor = [Define kColorBlack];
    leftView.font = [UIFont systemFontOfSize:22.0];
    leftView.text = @"¥";
    self.textMoney.leftView = leftView;
    self.textMoney.leftViewMode = UITextFieldViewModeAlways;
    
    UIView *lineView = [UIView new];
    lineView.backgroundColor = [Define kColorLine];
    [self.viewCurrent addSubview:lineView];
    [lineView mas_updateConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(0.5);
        make.left.right.equalTo(weakSelf.viewCurrent);
        make.top.equalTo(weakSelf.textMoney.mas_bottom).offset(-0.5);
    }];
    
    _btnWechat = [UIButton buttonWithType:UIButtonTypeCustom];
    _btnWechat.backgroundColor = [UIColor clearColor];
    _btnWechat.contentHorizontalAlignment = UIControlContentHorizontalAlignmentLeft;
    [_btnWechat.titleLabel setFont:[UIFont systemFontOfSize:16.0]];
    [_btnWechat setTitleColor:[Define kColorGray] forState:UIControlStateNormal];
    _btnWechat.adjustsImageWhenHighlighted = false;
    _btnWechat.titleEdgeInsets = UIEdgeInsetsMake(0, 18.0, 0, 0);
    [_btnWechat setTitle:@"微信支付" forState:UIControlStateNormal];
    [_btnWechat setImage:[UIImage imageNamed:@"choose_btn"] forState:UIControlStateNormal];
    [self.viewCurrent addSubview:self.btnWechat];
    [self.btnWechat mas_updateConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(weakSelf.viewCurrent).offset(35.0);
        make.top.equalTo(weakSelf.textMoney.mas_bottom).offset(30.0);
        make.width.equalTo(weakSelf.viewCurrent).offset(-70.0);
        make.height.mas_equalTo(20.0);
    }];
    
    CGRect rect_submit = CGRectMake(0, 0, [Define screenWidth] - 70.0, 44.0);
    _btnSubmit = [UIButton buttonWithType:UIButtonTypeCustom];
    [_btnSubmit setBackgroundImage:[UIImage imageWithDraw:[Define kColorCyanOff] sizeMake:rect_submit] forState:UIControlStateNormal];
    [_btnSubmit setBackgroundImage:[UIImage imageWithDraw:[Define kColorCyanOn] sizeMake:rect_submit] forState:UIControlStateHighlighted];
    [_btnSubmit.titleLabel setFont:[UIFont systemFontOfSize:20.0]];
    [_btnSubmit setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    _btnSubmit.layer.cornerRadius = 22.0;
    _btnSubmit.layer.masksToBounds = true;
    [_btnSubmit setTitle:@"去充值" forState:UIControlStateNormal];
    [_btnSubmit addTarget:self action:@selector(clickWithSubmit:) forControlEvents:UIControlEventTouchUpInside];
    [self.viewCurrent addSubview:self.btnSubmit];
    [self.btnSubmit mas_updateConstraints:^(MASConstraintMaker *make) {
        make.size.mas_equalTo(rect_submit.size);
        make.centerX.equalTo(self.view);
        make.top.equalTo(weakSelf.btnWechat.mas_bottom).offset(40.0);
    }];
    
}

- (void)clickWithSubmit:(UIButton *)sender {
    CGFloat money = [self.textMoney.text floatValue];
    if (money <= 0.0) {
        [Tool showHUDTipWithTipStr:@"请输入充值金额!"];
        return;
    }
    [self topUpMoney];
}

#pragma mark - UITextFieldDelegate
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return YES;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.textMoney resignFirstResponder];
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    
    NSString *tempText = [NSString stringWithFormat:@"%@%@",textField.text, string];
    tempText = [tempText stringByReplacingCharactersInRange:range withString:@""];
    NSArray *arrayCount = [tempText componentsSeparatedByString:@"."];

    if (arrayCount.count > 2) {
        return false;
    } else if ([tempText floatValue] > 5000.0) {
       textField.text = @"5000.0";
        return false;
    } else if ([tempText rangeOfString:@"."].location != NSNotFound) {
        //只能有两位小数
        NSRange tempRange = [tempText rangeOfString:@"."];
        if (tempText.length - tempRange.location - tempRange.length > 2) {
            return false;
        }
    }
    return  YES;
}

#pragma mark - WXApiEngineDelegate
- (void)shareEngineWXApi:(SendAuthResp *)response {
    if (response.errCode == -2) {
        //用户取消
        [Tool showHUDTipWithTipStr:@"您已取消授权！"];
    } else if (response.errCode == -4) {
        //用户拒绝授权
        [Tool showHUDTipWithTipStr:@"您已拒绝授权！"];
    } else {
        // 0(用户同意)
        _weixin_code = response.code;
        [self queryByAppOpenId];
    }
}

- (void)shareEnginePayment:(PayResp *)response {
    if (response.errCode == WXSuccess) {
        self.textMoney.text = @"";
        [Tool showHUDTipWithTipStr:@"充值成功！"];
    } else {
        NSString *strMsg = [NSString stringWithFormat:@"支付结果：失败！retcode = %d, retstr = %@", response.errCode,response.errStr];
        [Tool showHUDTipWithTipStr:strMsg];
    }
}

#pragma mark - Network
- (void)queryByAppOpenId {
    if ([NSString isBlankString:self.weixin_code]) {
        return;
    }
    WEAK_SELF
    [HttpAPIClient APIWeChatToCode:self.weixin_code Success:^(id result) {
        DebugLog(@"_________resutl:%@",result);
        NSString *errmsg = result[@"errmsg"];
        if (![NSString isBlankString:errmsg]) {
            [Tool showHUDTipWithTipStr:errmsg];
        } else {
            weakSelf.weixin_appOpenId = [NSString stringWithFormat:@"%@",result[@"openid"]];
            weakSelf.weixin_unionId = [NSString stringWithFormat:@"%@",result[@"unionid"]];
            [weakSelf bandWeixin];
        }
    } Failed:^(NSError *error) {
        
    }];
}

- (void)bandWeixin {
    if ([NSString isBlankString:self.weixin_appOpenId] || [NSString isBlankString:self.weixin_unionId]) {
        return;
    }
    UserManager *user = [UserManager getUser];
    NSString *user_id = [NSString stringWithFormat:@"%@",user.userId];
    NSString *union_id = [NSString stringWithFormat:@"%@",self.weixin_unionId];
    
    NSDictionary *param = @{@"userId":user_id,@"unionId":union_id,@"appOpenId":self.weixin_appOpenId};
    WEAK_SELF
    [HttpAPIClient APIClientPOST:@"bindWeixin" params:param Success:^(id result) {
        DebugLog(@"________result:%@",result);
        NSDictionary *data = [result[@"data"] mj_JSONObject][0];
        if ([data[@"ret"] integerValue] == 0) {
            [weakSelf topUpMoney];
        } else {
            [Tool showHUDTipWithTipStr:data[@"desc"]];
        }
    } Failed:^(NSError *error) {
        DebugLog(@"________error:%@",error);
    }];
}

/** 充值金额 */
- (void)topUpMoney {
   
    NSString *str_ip = [NSString getIPAddress];
    UserManager *info = [UserManager getUser];
    NSString *str_userId = [NSString stringWithFormat:@"%@",info.userId];
    NSString *str_amount = [NSString stringWithFormat:@"%.2f",[self.textMoney.text floatValue]];
    
    NSDictionary *param = @{@"userId":str_userId,@"amount":str_amount,@"ip":str_ip};
    
    DebugLog(@"_____________param:%@",param);
    [HttpAPIClient APIClientPOST:@"recharge" params:param Success:^(id result) {
        DebugLog(@"_______result:%@",result);
        NSDictionary *data = [result[@"data"] mj_JSONObject][0];
        if ([data[@"ret"] integerValue] == 1) {
            UIAlertController *alertControl = [UIAlertController alertControllerWithTitle:@"如需充值，请先绑定微信！" message:nil preferredStyle:UIAlertControllerStyleAlert];
            [alertControl addAction:[UIAlertAction actionWithTitle:@"确定" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                
                [[WMShareEngine sharedInstance] sendAuthRequest:self];
            }]];
            [alertControl addAction:[UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            }]];
            [self presentViewController:alertControl animated:YES completion:nil];
        } else if ([data[@"ret"] integerValue] == 0) {
            [[WMShareEngine sharedInstance] sendWeChatPaymentInfo:data[@"info"]];
        } else {
            [Tool showHUDTipWithTipStr:data[@"desc"]];
        }
    } Failed:^(NSError *error) {
        DebugLog(@"________error:%@",error);
    }];
}

@end

File: WhipMe/Controller/Member/TopUpViewController.swift
//
//  TopUpViewController.swift
//  WhipMe
//
//  Created by yangg on 2016/10/5.
//  Copyright © 2016年 -. All rights reserved.
//

import UIKit

class TopUpViewController: UIViewController, UITextFieldDelegate {

    private var textMoney: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.title = "充值"
        self.view.backgroundColor = Define.kColorBackGround()

        setup()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    func setup() {
        let viewCurrent = UIView.init()
        viewCurrent.backgroundColor = UIColor.white
        viewCurrent.layer.cornerRadius = 4.0
        viewCurrent.layer.masksToBounds = true
        self.view.addSubview(viewCurrent)
        viewCurrent.snp.updateConstraints { (make) in
            make.left.top.equalTo(10.0)
            make.right.equalTo(-10.0)
            make.height.equalTo(290.0)
        }
        
        let lblMoneyTitle = UILabel.init()
        lblMoneyTitle.backgroundColor = UIColor.clear
        lblMoneyTitle.textAlignment = NSTextAlignment.left;
        lblMoneyTitle.textColor = Define.kColorGray()
        lblMoneyTitle.font = KContentFont
        lblMoneyTitle.text = "充值金额"
        viewCurrent.addSubview(lblMoneyTitle)
        lblMoneyTitle.snp.updateConstraints { (make) in
            make.height.equalTo(40.0)
            make.width.equalTo(viewCurrent).offset(-60.0)
            make.left.equalTo(30.0)
            make.top.equalTo(viewCurrent)
        }
        
        textMoney = UITextField.init()
        textMoney.contentVerticalAlignment = UIControlContentVerticalAlignment.center
        textMoney.backgroundColor = UIColor.clear
        textMoney.textAlignment = NSTextAlignment.left
        textMoney.font = UIFont.systemFont(ofSize: 22.0)
        textMoney.keyboardType = UIKeyboardType.decimalPad
        textMoney.textColor = Define.kColorBlack()
        textMoney.delegate = self
        textMoney.placeholder = "请输入充值金额"
//        textMoney.attributedPlaceholder = NSMutableAttributedString.init(string: "请输入充值金额", attributes: [NSFontAttributeName:UIFont.systemFont(ofSize: 20.0), NSForegroundColorAttributeName:Define.kColorGray()])
        viewCurrent.addSubview(textMoney)
        textMoney.snp.makeConstraints { (make) in
            make.height.equalTo(50.0)
            make.width.equalTo(lblMoneyTitle.snp.width)
            make.left.equalTo(lblMoneyTitle.snp.left)
            make.top.equalTo(lblMoneyTitle.snp.bottom)
        }
        
        let leftView = UILabel.init(frame: CGRect.init(x: 0, y: 0, width: 40.0, height: 50.0))
        leftView.textAlignment = NSTextAlignment.center
        leftView.textColor = Define.kColorBlack()
        leftView.font = UIFont.systemFont(ofSize: 22.0)
        leftView.text = "¥"
        textMoney.leftView = leftView
        textMoney.leftViewMode = UITextFieldViewMode.always
        
        let lineView = UIView.init()
        lineView.backgroundColor = Define.kColorLine()
        viewCurrent.addSubview(lineView)
        lineView.snp.updateConstraints { (make) in
            make.height.equalTo(0.5)
            make.left.right.equalTo(viewCurrent)
            make.top.equalTo(textMoney.snp.bottom).offset(-0.5)
        }
        
        let btnWechat = UIButton.init(type: UIButtonType.custom)
        btnWechat.backgroundColor = UIColor.clear
        btnWechat.contentHorizontalAlignment = UIControlContentHorizontalAlignment.left
        btnWechat.titleLabel?.font = KContentFont
        btnWechat.setTitleColor(Define.kColorGray(), for: UIControlState.normal)
        btnWechat.adjustsImageWhenHighlighted = false
        btnWechat.titleEdgeInsets = UIEdgeInsets.init(top: 0, left: 18.0, bottom: 0, right: 0)
        btnWechat.setTitle("微信支付", for: UIControlState.normal)
        btnWechat.setImage(UIImage.init(named: "choose_btn"), for: UIControlState.normal)
        viewCurrent.addSubview(btnWechat)
        btnWechat.snp.updateConstraints { (make) in
            make.left.equalTo(35.0)
            make.top.equalTo(textMoney.snp.bottom).offset(30.0)
            make.width.equalTo(viewCurrent).offset(-70.0)
            make.height.equalTo(20.0)
        }
        
        let rect_submit = CGRect.init(x: 0, y: 0, width: Define.screenWidth() - 70.0, height: 44.0)
        let btnSubmit = UIButton.init(type: UIButtonType.custom)
        btnSubmit.setBackgroundImage(UIImage.imageWithDraw(Define.kColorCyanOff(), sizeMake: rect_submit), for: UIControlState.normal)
        btnSubmit.setBackgroundImage(UIImage.imageWithDraw(Define.kColorCyanOn(), sizeMake: rect_submit), for: UIControlState.highlighted)
        btnSubmit.titleLabel?.font = UIFont.systemFont(ofSize: 20.0)
        btnSubmit.setTitleColor(UIColor.white, for: UIControlState.normal)
        btnSubmit.layer.cornerRadius = 22.0
        btnSubmit.layer.masksToBounds = true
        btnSubmit.setTitle("去充值", for: UIControlState.normal)
        btnSubmit.addTarget(self, action: #selector(clickWithSubmit), for: UIControlEvents.touchUpInside)
        viewCurrent.addSubview(btnSubmit)
        btnSubmit.snp.makeConstraints { (make) in
            make.size.equalTo(rect_submit.size)
            make.centerX.equalTo(self.view)
            make.top.equalTo(btnWechat.snp.bottom).offset(40.0)
        }
        
    }
    
    func clickWithSubmit() {
        
    }
    

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.view.endEditing(false)
    }
    
    // MARK: - UITextFieldDelegate
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        
        let string_ns = textField.text as NSString?
        let string_temp = string_ns?.replacingCharacters(in: range, with: string)
        if string_temp?.characters.count == 0 {
            return true
        }
        
        let string_float:Float = Float(string_temp!)!
        print("text delegate is string_float:\(string_float)")
        
        if string_float > 5000.0 {
            textField.text = "5000"
            return false
        }
        
        
        
//        if (textField.tag == KTextTag)
//        {
//            NSString *tempText = [NSString stringWithFormat:@"%@%@",textField.text, string];
//            tempText = [tempText stringByReplacingCharactersInRange:range withString:@""];
//            NSArray *arrayCount = [tempText componentsSeparatedByString:@"."];
//            
//            if (arrayCount.count > 2)
//            {
//                return false;
//            } else if ([tempText floatValue] > _paidMoney) {
//                [self showisMessage:[NSString stringWithFormat:@"您的消费金额不能大于%.2f元！",_paidMoney]];
//                return false;
//            } else if ([tempText rangeOfString:@"."].location != NSNotFound) {
//                //只能有两位小数
//                NSRange tempRange = [tempText rangeOfString:@"."];
//                if (tempText.length - tempRange.location - tempRange.length > 2)
//                {
//                    return false;
//                }
//            }
//        }
        
        
        return true
    }
    
  

}
